applications/MeshMovingApplication/custom_utilities/parametric_linear_transform.h

```diff
@@ -0,0 +1,170 @@
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(ParametricLinearTransform);
+
+    using FunctionType = GenericFunctionUtility;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+    
+    /** Construct via axis & angle
+     *  @param rAxis axis of rotation (array of size 3)
+     *  @param rAngle angle of rotation (in radians, scalar)
+     *  @param rReferencePoint a point on the axis of rotation (array of size 3)
+     *  @param rTranslationVector translation vector (array of size 3)
+     */
+    ParametricLinearTransform(const Parameters& rAxis,
+                              const Parameters& rAngle,
+                              const Parameters& rReferencePoint,
+                              const Parameters& rTranslationVector);
+
+    /** Construct via euler angles
+     *  @param rEulerAngles euler angles (radians, array of size 3)
+     *  @param rReferencePoint origin of rotation (array of size 3)
+     *  @param rTranslationVector translation vector (array of size 3)
+     *  @note The euler angles follow the convention specified by @ref{Quaternion} (Z, -X', Z")
+     */
+    ParametricLinearTransform(const Parameters& rEulerAngles,
+                              const Parameters& rReferencePoint,
+                              const Parameters& rTranslationVector);
+
+    ParametricLinearTransform(const ParametricLinearTransform& rOther) = default;
+
+    ParametricLinearTransform(ParametricLinearTransform&& rOther) = default;
+
+    ParametricLinearTransform& operator=(const ParametricLinearTransform& rOther) = default;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /** Evaluate the transformation at the given location and apply it to a position vector, returning its transformed version
+     *  @note the transformation is not updated unless at least one transformation-related
+     *  quantity has changed since the last call to ParametricLinearTransform::Apply.
+     *  @param rPoint point to be transformed; it also acts as the current position input to @ref{GenericFunctionUtility}
+     *  @param t time, see @ref{GenericFunctionUtility}
+     *  @param X initial x-coordinate, see @ref{GenericFunctionUtility}
+     *  @param Y initial y-coordinate, see @ref{GenericFunctionUtility}
+     *  @param Z initial z-coordinate, see @ref{GenericFunctionUtility}
+     */
+    array_1d<double,3> Apply(const array_1d<double,3>& rPoint,
+                             const double t,
+                             const double X = 0.0,
+                             const double Y = 0.0,
+                             const double Z = 0.0);
+
+    ///@}
+
+private:
+
+    ///@name Private type definitions
+    ///@{
+
+    // Necessary to parse constants
+    static std::string ExtractFunctionBody(const Parameters& rParameters);
+
+    /// Class for storing and evaluating an array of @ref{GenericFunctionUtility}
+    template <std::size_t ArraySize>
+    class VectorFunction : public std::array<FunctionType::Pointer,ArraySize>
```

Comment:
afaik it is not a good idea to derive the STL types
Can you comment please?

Reply:
Well, ```VectorFunction``` lives in a closed environment and doesn't add any member variables to ```std::array```. That said, ```public``` inheritance is definitely not justified. Are you okay with a ```private``` inheritance?

Reply:
I am fine either way, I just wanted to ask if you have good reasons :)
(My knowledge abt this doesn't go beyond reading "don't derive the stl types" online ...)

Reply:
Here's what I understand: the dangers I know of are all tied to the stl destructors not being virtual. Since ```std::array``` is basically just a wrapper for a C array, nothing needs to be freed upon destruction, so the only issue we're left with is casting. 

If someone casts a ```VectorFunction``` to its base ```std::array``` somewhere, we end up with UB upon destruction because ```VectorFunction::~VectorFunction``` would be called on an ```std::array```. However, if the inheritance is ```private```, such a cast can never occur because the relationship between ```VectorFunction``` and ```std::array``` is hidden.

I could've just implemented ```VectorFunction``` with an ```std::array``` member, but that would've added an extra dereferencing every time ```VectorFunction``` got evaluated (not a huge deal, but why the extra overhead).

Reply:
ok for me, thx for the explanation/discussion